RFC: add mingw to appveyor matrix

tkelman opened the pull request:
As discussed here https://github.com/libgit2/libgit2/pull/2414#issuecomment-76848152

This is just a start, ~~and is only using the 32-bit mingw.org that is already installed on the AppVeyor VM's. I can keep working on this and add 32 and 64 bit MinGW-w64 distributions. Just need to pick which distribution to download.~~

It looks like there might be a build issue with mingw.org at the moment. On my fork I got
```
[ 16%] Building C object CMakeFiles/git2.dir/src/trace.c.obj
In file included from c:/projects/libgit2/src/common.h:66:0,
                 from c:/projects/libgit2/src/buffer.h:10,
                 from c:/projects/libgit2/src/trace.c:8:
c:/projects/libgit2/src/trace.c: In function 'git_trace_set':
 
c:/projects/libgit2/src/thread-utils.h:279:43: error: expected expression before ')' token
 # define GIT_MEMORY_BARRIER MemoryBarrier()
                                           ^
c:/projects/libgit2/src/trace.c:27:2: note: in expansion of macro 'GIT_MEMORY_BARRIER'
  GIT_MEMORY_BARRIER;
  ^
 
make[2]: 
*** [CMakeFiles/git2.dir/src/trace.c.obj] Error 1
 
make[1]: 
*** [CMakeFiles/git2.dir/all] Error 2
```
Which is interesting, since the cross-compile on Travis appears to be working.

tkelman commented:
Sorry for eating up a lot of queue time experimenting with things here. If I knew of a way to only run these on AppVeyor and not Travis, I would. Please cancel any redundant pending builds for this so other things can go through. I have some code in here to fail immediately for obsolete builds on AppVeyor when a newer build is pending for the same PR, this could probably be translated to bash and used with Travis too but I don't know what to query from their API.

The 32 and 64 bit MinGW-w64 builds both looked successful, but powershell may have been treating all the compiler warnings as errors or something and it wasn't running the tests. I'm trying to run the build steps from cmd instead of powershell, though that is being pickier about multiline commands.

tkelman commented:
I think I see why the cross-compile was working with `i586-mingw32msvc-gcc` but building with mingw.org on Windows isn't: the former is not setting `-DTHREADSAFE=ON`.

Anyway, this works and runs the tests now, at least for mingw-w64. There are a whole bunch of compiler warnings and 12 test failures due to building without openssl (I could go find a mingw-w64 openssl download), 4 "Unsupported URL protocol," and something else wrong with `online::clone::can_cancel` and `online::clone::certificate_invalid`. Could put these in `allow_failures:` to start with.

ethomson commented:
Wow!!!  Very cool.

With regards to the failing tests, we either need to build against openssl, or we need to simply not run the online tests.  (It should be enough to drop the `-ionline` argument from clar.)

tkelman commented:
I can get a cross-compiled binary of openssl from the opensuse build service. It requires some messy rpm repodata xml parsing, I have a version written in bash (filed prominently under things you shouldn't write in bash) that relies on `xmllint` which isn't installed on appveyor. I also have a partially complete version written in powershell, I got stuck at needing `sort -V` though (rpm-style version number sorting).

> (It should be enough to drop the `-ionline` argument from clar.)

Is that easily doable through `ctest` too? What would be preferred to start with, running fewer tests with mingw, or allowing failures?

tkelman commented:
I found an OpenSSL binary and got it installed, it built okay but gave `error -1 - failed to create ssl object` on i686 and segfaulted on x86_64. The binary was probably built with MSVC, but I likely need one built with the same MinGW variant I'm using. @akikoskinen where were you getting OpenSSL from in your builds?

carlosmn commented:
I don't think we should bother too much with OpenSSL on Windows but instead try to get the PR which makes us use WinHTTP from the mingw environments in working order.

ethomson commented:
@tkelman My thinking here is that allowed failures vs online tests are a little bit orthogonal.  And I agree with @carlosmn that we should work on getting the WinHTTP changes merged.  Until that point, I think that we shouldn't spend much time on OpenSSL, because therein probably lies madness.  :)

I think that what I would like to see is:  
* Don't run the `online` tests for now, as they assume HTTPS pretty much throughout.  I *think* that the easiest way to do that is just to not run `ctest`, but to run `libgit2_clar` instead (perhaps with the `-v` flag).  I admit that the nice thing about `ctest` is that it knows where `libgit2_clar` got built, so if that's particularly handy, we could [change the test definition](https://github.com/libgit2/libgit2/blob/master/CMakeLists.txt#L496) to know whether it should run the online tests or not.  Maybe something like:

    IF(WINHTTP OR OPENSSL_FOUND)
        ADD_TEST(libgit2_clar libgit2_clar -ionline)
    ELSE()
        ADD_TEST(libgit2_clar libgit2_clar -v)
    ENDIF()

(Note, not tested :smile: )

* We leave mingw as an allowed failure *for now*.  I think that phasing this in gradually would be nice.  For now the feedback will be helpful, and we can make sure that Important Milestones build on mingw.  If these prove to not be brittle, or if we have somebody bullish on mingw (and this person might be you?) then we should just flip them over to not allow failures.

* We should work towards WinHTTP on mingw.  I think that this is outside of the scope of this PR, but that this PR helps get us there with some nice feedback.  Once that lands we can turn `-ionline` back on here.

Thanks again, I'm very excited to see this.

akikoskinen commented:
My MinGW installation is on Fedora Linux. Currently I have Fedora 20 and Fedora 21 distros although their MinGW versions are pretty much on the same level with each other I think. There is OpenSSL available right from the official repositories.

I don't dare to try MinGW on Windows - it's such a can of worms that I would just get miserable :(

tkelman commented:
Thanks all for the input. @akikoskinen I also prefer cross-compiling, I like openSUSE since their build service is super easy to use, but the MinGW package selection is pretty similar between Fedora and openSUSE.

I'll rebase out the openssl-related commit and try out some modifications to the test setup as suggested by @ethomson. It looks like the new tests from #2932 are failing, I may try tweaking one of those ifdefs to see if mingw-w64 has `_chsize_s` available.

tkelman commented:
Looks like mingw-w64 has `_chsize_s`, but that doesn't fix the ftruncate test failure. I was hoping to only need to allow failures for mingw.org, but not quite yet I guess. Anyway, I think this could be merged in its current state if people like it.

tkelman commented:
> Looks like mingw-w64 has `_chsize_s`, but that doesn't fix the ftruncate test failure.

Silly me, I had a typo in my ifdef. Fixed now, ftruncate test passing on mingw-w64.

Travis has a big backlog of all the times I rebased and force-pushed this branch.  ba6c53b91b1d9c9dc49b7c0d8e8efbd9842c31dc is intended to work through that backlog a little quicker, but could be taken out since I'll admit I haven't tested that code outside of this PR.

tkelman commented:
:cake: I fixed the mingw.org build problem too. It was due to empty/missing defines in some of the trace code, and a separate problem with `VOLUME_NAME_DOS` in a test - neither the trace code nor the tests are built by the Travis mingw cross-compile. These could be fixed by changing `_WIN32_WINNT` to at least `0x0600`, or adding a couple of defines (only under the conditions where they'd otherwise be missing) to `mingw-compat.h`. I went with the latter for now, felt less invasive.

3 test failures on mingw.org, but could be much worse given the toolchain.

ethomson commented:
Thanks for all the hard work here @tkelman - I'll dig into this when I have a moment.

Re cross-compiling - I agree that mingw on Windows is a challenge.  However WINE introduces a handful of other issues - we don't work under WINE at the moment (or we didn't the last time that anybody I know tried it) and I don't want to end up testing WINE instead of us, as I'm sure you can sympathize with.

Thanks again, I know you've put a ton of effort into this.

ethomson commented on src/win32/posix_w32.c:
```diff
@@ -55,7 +55,7 @@ int p_ftruncate(int fd, git_off_t size)
 	}
 
-#if !defined(__MINGW32__)
+#if !defined(__MINGW32__) || defined(__MINGW64_VERSION_MAJOR)
```
A recent PR suggests that [`MINGW_HAS_SECURE_API`](https://github.com/karvaporo/libgit2/commit/73682705093e028af823759b1ede5c9f1ec029fc) might be defined when this is supported.  This might be more idiomatic and future-proof.

ethomson commented:
This looks really nice to me.  Not sure if that `MINGW_HAS_SECURE_API` definition is helpful or not, I'll let you decide what you want to do with that information.

I know that @phkelley has played in the mingw space, I don't know if he has any additional thoughts.

tkelman commented:
> Re cross-compiling - I agree that mingw on Windows is a challenge. However WINE introduces a handful of other issues - we don't work under WINE at the moment (or we didn't the last time that anybody I know tried it) and I don't want to end up testing WINE instead of us, as I'm sure you can sympathize with.

I agree there. I should've been a little more specific. I like cross-compiling for building binaries, since it's more easily automatable and repeatable via Docker, OBS, Koji, what have you - native dependency management on Windows is still an absolute mess, letting a Linux package manager do the work for you actually works really well for getting binaries built. Testing those cross-compiled binaries via Wine is good to do when things are simple, but also fraught with Wine quirks when things aren't simple. As an anecdote, Julia has a rewritten garbage collector as of a few months ago, which mysteriously broke the step in cross-compilation that depended on Wine - probably due to a bug in Wine.

> Not sure if that `MINGW_HAS_SECURE_API` definition is helpful or not, I'll let you decide what you want to do with that information.

I'll push a test commit and see whether that does the job as well as my initial version. If so, I can rearrange things a little.

tkelman commented:
> I'll push a test commit and see whether that does the job as well as my initial version. If so, I can rearrange things a little.

Looks like it worked fine. I took out my versions of those commits and did it with a cherry-pick instead, so this vs #2959 could hopefully be merged in either order.

ethomson commented:
Thanks for the hard work here, @tkelman .  Really happy to see this done up so nicely.

tkelman commented:
Cool, you're very welcome. Happy to help make mingw(-w64) easier to support for you guys. Especially if https support without needing openssl can be tidied up and finalized by someone who knows what they're doing.

Don't hesitate to revert either https://github.com/tkelman/libgit2/commit/8008ab6a4158416d095212c9782fa1622ef994a5 or https://github.com/tkelman/libgit2/commit/ba6c53b91b1d9c9dc49b7c0d8e8efbd9842c31dc if they cause any trouble. They were good to have within the context of this PR, but could also be useful in doing the job of cutting down on queue time elsewhere.
